### ISSUE TITLE
NUnit3 console runner executes all SetUpFixtures before running tests in different namespaces. (change of behavior from 2.6.4 console runner)

### ISSUE THREAD

**Issue by cbchenoweth** [nunit/nunit#1330]
When running a suite of tests that cross multiple namespaces (that has a SetUpFixture in each namespace), 
all of the SetUpFixtures [OneTimeSetUp] methods get executed before all the tests.  Then all of the tests are run.  Then all the SetUpFixtures [OneTimeTearDown] functions are run.

This is a different behavior then what happened in the 2.6.4 console runner.
(before it would "group" the tests by namespace)

Say you have these classes in your tests:

```
//SetUpFixture in Namespace1
namespace Tests.Namespace1
{
    [SetUpFixture]
    public class Namespace1SetupAndTeardown
    {
        [OneTimeSetUp]
        public void BeforeNamespace1()
        {
        }

        [OneTimeTearDown]
        public void AfterNamespace1()
        {
        }
    }
}

//TestFixture in Namespace1
namespace Tests.Namespace1
{
    [TestFixture]
    public class Namespace1Tests
    {
        [Test]
        public void Test1()
        {
        }
    }
}

//SetUpFixture in Namespace2
namespace Tests.Namespace2
{
    [SetUpFixture]
    public class Namespace2SetupAndTeardown
    {
        [OneTimeSetUp]
        public void BeforeNamespace2()
        {
        }

        [OneTimeTearDown]
        public void AfterNamespace2()
        {
        }
    }
}

//TestFixture in Namespace2
namespace Tests.Namespace2
{
    [TestFixture]
    public class Namespace2Tests
    {
        [Test]
        public void Test2()
        {
        }
    }
}
```

**In the console runner for 2.6.4, you would see the functions executed in this order:**
Namespace1SetupAndTeardown.BeforeNamespace1()
Namespace1Tests.Test1()
Namespace1SetupAndTeardown.AfterNamespace1()

Namespace2SetupAndTeardown.BeforeNamespace2()
Namespace2Tests.Test2()
Namespace2SetupAndTeardown.AfterNamespace2()

**In the console runner for 3, I am seeing the functions get executed in this order:**
**(Note the different order the setups/teardowns are being executed)**
Namespace1SetupAndTeardown.BeforeNamespace1()
Namespace2SetupAndTeardown.BeforeNamespace2()

Namespace1Tests.Test1()
Namespace2Tests.Test2()

Namespace1SetupAndTeardown.AfterNamespace1()
Namespace2SetupAndTeardown.AfterNamespace2()

This is causing some of our test suites to fail, becuase the setup/teardowns are being "mixed" across namespaces.

Note(we are running tests single threaded, command line arguments "--workers=1 --agents=1"

**Would this be considered a bug?  Or is that behavior just "not supported"?**

**Comment by CharliePoole** [comment]
The order of execution is not deliberate, but it is an outcome of the order in which the work is added to the various queues. Initially, NUnit starts out with the test represented by your assembly as a whole. Usually, there is no code associated with this test, so it immediately adds all its child tests. This continues rapidly until we come to a test with real code to execute. In your case, this would be one of your namespace SetupFixtures. The setup fixture has to execute it's setup before the child items are added to the queue. As a result, it's likely that all the setup fixtures will be queued ahead of any of the component fixtures.

Note that none of this is planned or guaranteed. There is no order guarantee for a SetUpFixture except that it's OneTimeSetUp method will run to completion before any child tests are executed and it's OneTimeTearDown method will run after them.

As a workaround, run tests for now  using --workers=0. I believe this will (still accidentally) give you the same result as for V2. Please let us know if it works.

This looks to me like a case that requires __Exclusion Groups__ as defined at https://github.com/nunit/dev/wiki/Framework-Parallel-Test-Execution This feature is not yet implemented but is documented as issue #165

**Comment by cbchenoweth** [comment]
I am just splitting up the suits into separate ones, so that they run isolated from each other.
(don't want to "hack" around the problem.  Want to fix how we use it so that it will work as it is intended to.)

Thanks for looking over this with me.
